Skip deleting database servers on agent shutdown during binary upgrade

greedy52 opened the pull request:
Partial fix #20126 
- #20126

Problem:
During [Teleport binary upgrade process](https://goteleport.com/docs/management/admin/upgrading-the-teleport-binary/), a child db agent is forked and starts to proxy the same databases as the parent. However, when the parent is getting "closed", the parent deletes the DatabaseServer heartbeats it creates as part of the cleanup. This cleanup causes the heartbeats in the child process to go into a bad state. The heartbeats eventually fix themself in a few minutes but during that period, users cannot connect.

With this fix, the live db connections will still break during the binary upgrade as the parent quits. But db servers will **not** disappear and an immediate re-connect will **succeed**. e.g.
```
$ tsh db connect steve-rds --db-user alice
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 10001
Server version: 8.0.28 Source distribution

...
mysql> use test
Database changed
mysql> select * from users;
ERROR 1146 (42S02): Table 'test.users' doesn't exist

mysql> select * from users;
No connection. Trying to reconnect...
ERROR 2013 (HY000): Lost connection to MySQL server during query
Connection id:    10002
Current database: test

ERROR 1146 (42S02): Table 'test.users' doesn't exist
```

A new ticket #21158 is created for connection draining

greedy52 committed:
Skip deleting database servers on agent shutdown during upgrade

Tener reviewed (approved):
I wonder if, from a reliability and complexity perspective, it wouldn't be better to ensure that the forked process always creates fresh identifiers for itself to use.

smallinsky reviewed (commented):
Generally look good to me but wan to clarify why the `collectStatuses` call is needed. 

@espadolini I would be happy to get your feedback about this approach. 


@greedy52 Do you know if bug also occurs for other services like APP Access ?

espadolini commented:
I'd prefer something that also handles in-process restarts caused by CA rotations; we should be able to add some value to the shutdown context to know if we're shutting down because we've started a new `TeleportProcess` or a new unix process.

Whatever the approach, we will also be able to use it for app access (see #14422).

espadolini commented:
Untested toy implementation of my idea: fe39a15f99c637fed6d87855c88ac0b963de70a3

greedy52 commented:
> fe39a15

Cool! Let me play around with this. And I will do app access as well.

greedy52 commented on lib/service/signals.go:
```diff
@@ -591,3 +591,8 @@ func (process *TeleportProcess) getForkedPIDs() []int {
 }
+
+func (process *TeleportProcess) hasForkedProcess() bool {
+	process.collectStatuses()
```
No, it's not required. I thought a refresh is needed but `collectStatuses` is called by `syscall.SIGCHLD` already.

greedy52 commented:
> I wonder if, from a reliability and complexity perspective, it wouldn't be better to ensure that the forked process always creates fresh identifiers for itself to use.

Currently `DatabaseServer` is keyed by
```
        Key: backend.Key(dbServersPrefix,                                
            server.GetNamespace(),                                
            server.GetHostID(),                                
            server.GetName()),      
```
`HostID` should be the same for parent/child process. Maybe we can add a PID to identify the process. That will simplify this case but might complicate other things along the way.

Tener commented:
> > I wonder if, from a reliability and complexity perspective, it wouldn't be better to ensure that the forked process always creates fresh identifiers for itself to use.
> 
> Currently `DatabaseServer` is keyed by
> 
> ```
>         Key: backend.Key(dbServersPrefix,                                
>             server.GetNamespace(),                                
>             server.GetHostID(),                                
>             server.GetName()),      
> ```
> 
> `HostID` should be the same for parent/child process. Maybe we can add a PID to identify the process. That will simplify this case but might complicate other things along the way.

I like this idea quite a lot, but I concur this may have unforeseen consequences.

espadolini commented:
Picking a random uuid in `db.init` and using `<host id>_<random uuid>` for the "host ID" in backend keys for `database_server` and `app_server` would work, but the problem with that is that `Upsert{Application,Database}Server` only takes a resource and constructs the backend key based on it; I'd be +1 for adding an "instance ID" to the heartbeat itself (it would help with debugging too, if you're seeing weird things and you notice that instance IDs keep changing it would mean that the agent is repeatedly restarting, for example), but that seems way more involved of a change.

greedy52 committed:
implement Shutdown(context.Context)

greedy52 commented:
I've updated the change with the **context** approach and added a `Shutdown` for app/db server. The context is propagated to the shutdown function and the server will decide what to do with it. Let me know how it looks. Thanks!

Kube server may have the same problem but a weird thing service is currently calling `http.Server.Shutdown` at the moment. Likely a more problematic bug.
https://github.com/gravitational/teleport/blob/7fab8fadc6beaa3dc872540e1a24e1ce7ce6a664/lib/service/kubernetes.go#L286

smallinsky commented on lib/services/process.go:
```diff
@@ -0,0 +1,54 @@
+	// will take over the heartbeats so do NOT delete them in that case. In
+	// worst case scenarios if the child fails to register new heartbeats, the
+	// old ones will get deleted automatically upon expiry.
+	if HasProcessForked(ctx) {
+		return false
+	}
+	return true
```
```suggestion
	return !HasProcessForked(ctx)
```

greedy52 committed:
and dedicated "XXXAndDelete" functions

greedy52 commented on lib/srv/db/server.go:
```diff
@@ -521,7 +521,7 @@ func (s *Server) unregisterDatabase(ctx context.Context, database types.Database
 	}
 	// Stop heartbeat, labels, etc.
-	if err := s.stopProxyingDatabase(ctx, database); err != nil {
+	if err := s.stopProxyingDatabase(ctx, database, true); err != nil {
```
Made two separate versions like  `stopProxyingDatabase` and `stopProxyingAndDeleteDatabase`. It's a bit more code/work but it's definitely more readable

espadolini commented on lib/service/service.go:
```diff
@@ -4798,6 +4803,11 @@ func (process *TeleportProcess) StartShutdown(ctx context.Context) context.Conte
 	warnOnErr(process.stopListeners(), process.log)
 
+	// populate context values
+	if len(process.getForkedPIDs()) > 0 {
```
This only catches the across-process restart on SIGHUP, but doesn't catch the in-process restart on CA rotations; instead of checking this here, can we just inject the `ProcessForkedContext` (which should be `ProcessRestartingContext`, maybe) in the two places in which we call `StartShutdown` for restarts?

greedy52 replied:
> This only catches the across-process restart on SIGHUP

The [upgrade guide](https://goteleport.com/docs/management/admin/upgrading-the-teleport-binary/) instructs a `SIGUSR2`, then followed by a `SIGQUIT` when the child is up. So I was thinking the in-process restart can be added separately with a different context value.

Though now you mention it, it does feel easier to consolidate them into one. Is this the only place for in-process restart?
https://github.com/gravitational/teleport/blob/2fe049e6c3d0b1438eef6b0d6493fef3ee1f2bd3/lib/service/service.go#L735-L738

espadolini replied:
You're actually right, we ought to check both possibilities, as the shutdown on SIGQUIT after a SIGUSR2 would just be a normal shutdown (not that SIGUSR2 can ever be used in a sensible way, mind you).

I believe that's the only place where we do in-process restarts. Keeping them separate would help because eventually we should get rid of in-process restarts altogether (host CA rotations/credentials reissuance is the only reason we still have that, but it should be possible to swap out credentials on the fly) and only SIGHUP/SIGUSR2 would remain.

greedy52 replied:
cool. then I will merge this PR as it is and I will do a separate PR for the in-process restart and test out the rotation scenario for #4422

greedy52 committed:
Merge branch 'master' into STeve/20126_db_agent_upgrade
